WooCommerce/Classes/Authentication/SiteCredentialLoginUseCase.swift

```diff
@@ -74,65 +75,87 @@ private extension SiteCredentialLoginUseCase {
         }
     }
 
-    func loginAndAttemptFetchingJetpackPluginDetails(username: String, password: String) {
-        handleCookieAuthentication(username: username, password: password)
-        retrieveJetpackPluginDetails()
-    }
-
-    func handleCookieAuthentication(username: String, password: String) {
-        guard let loginURL = URL(string: siteURL + Constants.loginPath),
-              let adminURL = URL(string: siteURL + Constants.adminPath) else {
-            DDLogWarn("⚠️ Cannot construct login URL and admin URL for site \(siteURL)")
-            let error = NSError(domain: SiteCredentialLoginError.errorDomain, code: -1)
-            errorHandler?(.genericFailure(underlyingError: error))
+    func startLogin(username: String, password: String, onSuccess: @escaping () -> Void, onFailure: @escaping (SiteCredentialLoginError) -> Void) {
+        guard let request = buildLoginRequest(username: username, password: password) else {
+            DDLogError("⛔️ Error constructing login request")
             return
         }
-        // Prepares the authenticator with username and password
-        let config = CookieNonceAuthenticatorConfiguration(username: username,
-                                                           password: password,
-                                                           loginURL: loginURL,
-                                                           adminURL: adminURL)
-        let network = WordPressOrgNetwork(configuration: config)
-        let authenticationAction = JetpackConnectionAction.authenticate(siteURL: siteURL, network: network)
-        stores.dispatch(authenticationAction)
+        let session = URLSession(configuration: .default, delegate: self, delegateQueue: nil)
+        let task = session.dataTask(with: request) { data, response, error in
+            DispatchQueue.main.async {
+                if let error = error as? NSError {
+                    /// The request is manually cancelled in the session delegate
+                    /// when the login succeeds and redirects to the nonce retrieval URL,
+                    /// so we consider this a success.
+                    if error.domain == "NSURLErrorDomain", error.code == -999 {
+                        return onSuccess()
+                    }
+                    return onFailure(.genericFailure(underlyingError: error))
+                }
+                guard let response = response as? HTTPURLResponse else {
+                    return onFailure(.invalidLoginResponse)
+                }
+                if response.statusCode == 404 {
+                    return onFailure(.inaccessibleLoginPage)
+                }
+                if let data, let html = String(data: data, encoding: .utf8) {
+                    // scrape html
+                    if html.contains("<div id=\"login_error\">") {
+                        return onFailure(.wrongCredentials)
```

Comment:
Could we extract the content of the `div` here and present it as the error message directly? For example for the ReCaptcha plugin we tested, the error is still contained in the `login_error`, but it's not credentials issue, so it's still misleading to say it's.

Reply:
Yup, done with regex in the subsequent commits.